receiver/kymastatsreceiver/kyma_scraper.go

```diff
@@ -0,0 +1,208 @@
+}
+
+func (ks *kymaScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	stats, err := ks.collectModuleStats(ctx)
+	if err != nil {
+		return pmetric.Metrics{}, err
+	}
+
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	for _, s := range stats {
+		ks.mb.RecordKymaModuleStatusStateDataPoint(now, int64(1), s.state)
+		rb := ks.mb.NewResourceBuilder()
+		rb.SetK8sNamespaceName(s.namespace)
+		rb.SetKymaModuleName(s.kind)
+		for _, c := range s.conditions {
+			val := conditionStatusToValue(c.status)
+			ks.mb.RecordKymaModuleStatusConditionsDataPoint(now, val, c.reason, c.status, c.condType)
+		}
+		ks.mb.EmitForResource(metadata.WithResource(rb.Emit()))
+	}
+
+	return ks.mb.Emit(), nil
+}
+
+func (ks *kymaScraper) collectModuleStats(ctx context.Context) ([]moduleStats, error) {
+	var res []moduleStats
+	for _, gvr := range ks.moduleGVRs {
+		moduleList, err := ks.client.Resource(gvr).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			ks.logger.Error("Error fetching module list",
+				zap.Error(err),
+				zap.String("group", gvr.Group),
+				zap.String("version", gvr.Version),
+				zap.String("resource", gvr.Resource))
+			return nil, err
+		}
+
+		for _, module := range moduleList.Items {
+			stats, err := ks.unstructuredToStats(module)
+			if err != nil {
+				ks.logger.Error("Error converting unstructured module to stats",
+					zap.Error(err),
+					zap.String("name", module.GetName()),
+					zap.String("namespace", module.GetNamespace()),
+					zap.String("kind", module.GetKind()),
+				)
+				continue
+			}
+
+			res = append(res, *stats)
+
+			// Take only the first valid module custom resource
+			break
+		}
+	}
+
+	return res, nil
+}
+
+func (ks *kymaScraper) unstructuredToStats(module unstructured.Unstructured) (*moduleStats, error) {
+	status, found, err := unstructured.NestedMap(module.Object, "status")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, &fieldNotFoundError{"status"}
+	}
+
+	state, found, err := unstructured.NestedString(status, "state")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, &fieldNotFoundError{"state"}
+	}
+
+	unstructuredConds, found, err := unstructured.NestedSlice(status, "conditions")
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, &fieldNotFoundError{"conditions"}
+	}
+
+	stats := &moduleStats{
+		state:     state,
+		namespace: module.GetNamespace(),
+		kind:      module.GetKind(),
+	}
+
+	for _, unstructuredCond := range unstructuredConds {
+		cond, err := ks.unstructuredToCondition(unstructuredCond)
+		if err != nil {
+			return nil, err
+		}
+		stats.conditions = append(stats.conditions, *cond)
+	}
+
+	return stats, nil
+}
+
+func (ks *kymaScraper) unstructuredToCondition(cond any) (*condition, error) {
+	condMap, ok := cond.(map[string]any)
+	if !ok {
+		return nil, errors.New("condition is not a map")
+	}
+
+	condType, found, err := unstructured.NestedString(condMap, "type")
```

Comment:
all the following code is based on the expectation that the conditions implement fields similar to metav1.conditions. This is not mandatory. We should not fail if a module either does not implement conditions at all, or if those conditions do not follow the same patterns as metav1.conditions (even though it is recommended that conditions are based on metav1)

Reply:
We assume that the CR has `metav1.condition` implementation. When a CR has no `metav1.condition` implementation is not scrapeable

Reply:
so in the end, even though there is no real contract (and it would be quite simple for us to exclude the conditions for such modules) we just fail with scraping for those modules?

Reply:
In case a module doe not have conditions or is not following the conventions, we should just skip that part, the same way as we skip when we cannot find a resource for a module CRD